### PR TITLE
pgx: Make it actually use pgx

### PR DESCRIPTION
**Pull request by AmandaCameron**
This modifies the pgx driver to actually use pgx directly, and adds a `WithPgxInstance` function, which will let people who are using the pgx driver directly run migrations using the package as a library.

### CODE REVIEW

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:225]: We'll need to revisit this once the fields are renamed. The current field names make this logic hard to follow.

The `pgx.Conn` should be closed first and only released if the stdlib was used.

**Review by dhui — CHANGES_REQUESTED** [review]

**Comment by dhui** [comment]
Thanks for the PR!

File: database/pgx/pgx.go
Comment (AmandaCameron) [reply database/pgx/pgx.go:225]: I'm under the impression that if we got it from AquireConn we're just "borowing" it from the sql.DB, so we shouldn't be closing it, but passing it back to the sql.DB instance with ReleaseConn, which is why I did it this way.

[The docs for AcquireConn](https://pkg.go.dev/github.com/jackc/pgx/v4/stdlib#AcquireConn) say: 
> AcquireConn acquires a *pgx.Conn from database/sql connection pool. It must be released with ReleaseConn.

Comment (dhui) [reply database/pgx/pgx.go]: Ahhh, that makes sense. e.g. if the connection is create/acquired via `AcquireConn()`, it should be returned/released via `ReleaseConn()` and not close. Similarly, [`database.sql.Conn.Raw()`](https://github.com/golang/go/blob/go1.16.5/src/database/sql/sql.go#L1929) doesn't close connections and releases it when finished.

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:234]: There are 3 cases `Close()` needs to handle:
1. Created with `Open()`
    * `p.db` and `p.conn` are both non-`nil`
    * should release the connection and close the DB
2. Created with `WithInstance()`
    * `p.db` and `p.conn` are both non-`nil`
    * _Should_ release the connection but leave the DB open for the caller of `WithInstance()` to close. However, this isn't the current behavior which is to close `p.db`, and existing clients may expect the current behavior, so let's leave it as is. e.g. close both `p.db` and release `p.conn`.
3. Created with `WithConnection()`
    * `p.db` is `nil` and `p.conn` is non-`nil`
    * Should leave the connection for the caller of `WithConnection()` to release the connection.

Per this [comment](https://github.com/golang-migrate/migrate/pull/593#discussion_r662378216), I think we should be releasing connection and never closing it

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:61]: Just realized that changing this a from a `*sql.Conn` to a `*pgx.Conn` will make it harder to [factor out common db driver logic](https://github.com/golang-migrate/migrate/issues/549) between the pgx and postgres drivers...

I don't see any direct pgx functionality being used in this PR. Are there any plans to do so?
Can we go back to using `*sql.Conn`? We can use [`Raw()`](https://golang.org/pkg/database/sql/#Conn.Raw) or [`Conn()`](https://pkg.go.dev/github.com/jackc/pgx/v4@v4.11.0/stdlib#Conn.Conn) to get the underlying `*pgx.Conn` if necessary.

**Review by dhui — CHANGES_REQUESTED** [review]
Thanks for addressing the PR feedback!

File: database/pgx/pgx.go
Comment (AmandaCameron) [reply database/pgx/pgx.go:61]: The benefit of using the pgx.Conn over sql.Conn is you can get a pgx.Conn from a sql.DB -- but you can't get a sql.DB from a pgx.Conn. This is the issue I was running into with a program of mine that uses pgx directly, for some speed benefits and some benefits around de/serialisation that can't be done through `database/sql` 

That said, `lib/pg`'s README states that it's in maintenance mode, and to use pgx instead, and I'm not aware of any other popular postgres go libraries.

Comment (dhui) [reply database/pgx/pgx.go:61]: Can a `pgx.Conn` be converted to a `sql.Conn`? I didn't see a way in the [docs](https://pkg.go.dev/github.com/jackc/pgx/v4@v4.11.0/stdlib). In your project(s) can you get a `sql.Conn` even though you're using pgx?

> That said, `lib/pg`'s README states that it's in maintenance mode, and to use pgx instead, and I'm not aware of any other popular postgres go libraries.

`migrate` needs to continue supporting `lib/pq` since there are [still active users](https://github.com/lib/pq/issues/1010#issuecomment-846286396) even though it's in maintenance mode. There isn't much of a difference between the pgx and postgres drivers (diff the source code to see for yourself how trivial the differences are) and contributors shouldn't need to update both drivers when making changes.

Comment (AmandaCameron) [reply database/pgx/pgx.go:61]: The only way I could find to get a sql.Conn/DB in my program would be to open a separate connection using sql.Open instead of just passing in the pgx.Conn my program already created to do it's database operations, which just doesn't feel that clean to me.

Comment (dhui) [reply database/pgx/pgx.go:61]: Let's leave the pgx db driver as is for now until we have a usecase for using `pgx.Conn` directly. Hopefully, we'll have the postgres and pgx db drivers refactored by then... Thanks again for the PR!

Comment (AmandaCameron) [reply database/pgx/pgx.go:61]: Am I understanding correctly that you mean that we'll leave the driver as it, and not merge this? If so, I'm fine with the MR getting closed, since I realised sometime during this process that the driver doesn't even *have* to be in-tree, and I'm fine with maintaining my own seperate package with this MR's contents extracted out.

Comment (dhui) [reply database/pgx/pgx.go:61]: Yeah, let's leave the driver as is for now and not merge this PR.

Comment (treuherz) [reply database/pgx/pgx.go:61]: You can get an `*sql.DB` from a pgx connection by getting its `ConnConfig` with `Config()` and then passing that to [`stdlib.OpenDB()`](https://pkg.go.dev/github.com/jackc/pgx/v4@v4.13.0/stdlib#OpenDB). Would this provide a path forward for this PR? I know it doesn't seem as clean as being able to adapt `pgx.Conn` directly, but without it it just leaves that are using the pgx "native" interface to have to convert to use stdlib themselves

Comment (dhui) [reply database/pgx/pgx.go:61]: Won't `stdlib.OpenDB()` create a new connection instead of using the existing one?

Comment (treuherz) [reply database/pgx/pgx.go:61]: Yes but if a user who's using `pgx.Conn` everywhere else wants to call this library then they have to do that anyway

Comment (dhui) [reply database/pgx/pgx.go:61]: When a db driver is created with an existing connection, the user's [expectation](https://github.com/golang-migrate/migrate/issues/578) is for that connection to be used by `migrate` but managed by the user.

Comment (treuherz) [reply database/pgx/pgx.go:61]: Ok, I see your point there. 

I guess my qualm is just that, without this PR, pgx users that weren't otherwise using the stdlib compatibility layer need to import it and then open a connection with it themselves in ordert to pass it to `NewWithDatabaseInstance`. And if you're not using the `WithDatabaseIstance` entrypoint then it's still an awkward way in as afaict the URL-style connstrings aren't compatible due to `migrate` looking for a `pgx://` scheme instead of `postgres://`, which pgx itself doesn't do. If you're already bought in to pgx and try to pick up `migrate`, it's tricky to figure out how to get the puzzle pieces to slot together. I think pushing for a unified pgx/pq driver that uses the `sql.DB` interface will leave that usecase unserved.

Comment (justinclift) [reply database/pgx/pgx.go:61]: Ugh.  Just ran into this ourselves.  We're using pgx/v5 without the stdlib compat layer, and it seems like migrate doesn't really support using an existing pgx connection.

Not a deal breaker, but it would have been nice not having to duplicate a bunch of connection setup/handling code, just for the brief migrate connection. :wink:

**Comment by MattBrittan** [comment]
FYI - pgx release 5.5 has added [`pool.OpenDBFromPool`](https://pkg.go.dev/github.com/jackc/pgx/v5@v5.5.0/stdlib#OpenDBFromPool); not a perfect solution but it does avoid the need to open a separate connection (for those using `pgx.Pool`).

**Pull request closed without merging**
